rstudio/rstudio
# Bugfix/busy session async

There are still some TODO comments here that I'm going to go over with Jonathan. Any feedback on how to make this better is much appreciated :) 

### Intent

Supports new configuration options to allow immediate or delayed async rpc and handling of selected operations offline - so they run even when the main thread is busy. 

### Approach

There are four new configuration options to rsession.conf:

    # settings to keep connectivity alive
    session-async-rpc-enabled=1
    session-async-rpc-timeout-ms=0

    # settings to process requests offline
    session-handle-offline-enabled=1
    session-handle-offline-timeout-ms=500

When either timeout-ms option is 0, the HttpConnection main request thread is used to convert all rpc requests to async or handle requests offline immediately. When it is non zero, the new OfflineService polling thread will do so if it wakes up and sees that the R session is currently executing, and the request was received longer than the specified timeout.

Periodically, offline service emits memory changed events in this same loop since those are only dispatched from the main thread.

An isMainThread check is added to key methods in r::exec so that the system logs and error and avoids stack corruption if an offline operation tries to use R. The app changes required include isWebsiteProject() - checking if R markdown is installed and findProgram(..) that calls R's Sys.which. There are TODOs in the code about how improving my solution to these problems.

The async-rpc requests are disabled from client_init until the first get_events request is made since the client only expects regular responses during setup.

Added a true/false user preference "Session protocol debug". When true, it logs debug messages to the file ~/.local/share/rstudio/log/rsession-(user).log for requests received and events sent from the rsession process. 

### Automated Tests

TBD: I would like to script up some tests that will kick off a system("sleep 60"), then verify that some operations can still proceed, and that we can queue up lots of console input messages, then interrupt the sleep.  It would be great to run these tests against a variety of configurations. We can also setup a mode to run all of the automated tests with various configuration options so that we can get more complete coverage of these new code paths.

Add a test in which R code updates the system PATH to include an executable checked by the rsession's findProgramOnPath.

### QA Notes

The defaults should not change any behavior right now as these options are disabled by default. 

Enabling and setting either or both of the timeout-ms values to 0 will ensure all requests are handled offline immediately if they are offlineable. If not, they are converted to async. With timeout-ms > 0, the request flow will only be altered if the R main thread is kept occupied for longer than the timeout-ms value. In other words, you will only see async rpcs used, or requests handled offline intermittently.

Also notice that when handle-offline is enabled, memory changed events should continue to be delivered even if R is busy.

There are two utilities in the code that have been changed to support "offline" requests: rmarkdownAvailable() and findProgramOnPath(). The rmarkdownAvailable() flag is now cached at startup. The findProgramOnPath() now uses the rsession's path env var which on Windows might drift out of sync if R makes changes to the PATH on the fly. 

### Checklist

- [ x] If this PR adds or changes UI, the updated UI meets [accessibility standards](https://github.com/rstudio/rstudio/wiki/Accessibility)
- [ x] A reviewer is assigned to this PR (if unsure who to assign, check Area Owners list)
- [ x] This PR passes all local unit tests
